src/Terminal.php

```diff
@@ -4,163 +4,120 @@
 
 /**
  * @author Michael Woodward <mikeymike.mw@gmail.com>
+ * @author Aydin Hassan <aydin@hotmail.co.uk>
  */
-class Terminal implements TerminalInterface
+interface Terminal
 {
     /**
-     * @var bool
+     * Get the available width of the terminal
      */
-    private $isTTY;
+    public function getWidth() : int;
 
     /**
-     * @var bool
+     * Get the available height of the terminal
      */
-    private $isCanonical = false;
+    public function getHeight() : int;
 
     /**
-     * @var int
+     * Disables echoing every character back to the terminal. This means
+     * we do not have to clear the line when reading.
      */
-    private $width;
+    public function disableEchoBack() : void;
 
     /**
-     * @var int
+     * Enable echoing back every character input to the terminal.
      */
-    private $height;
+    public function enableEchoBack() : void;
 
     /**
-     * @var string
+     * Is echo back mode enabled
      */
-    private $details;
+    public function isEchoBack() : bool;
 
     /**
-     * @var string
+     * Disable canonical input (allow each key press for reading, rather than the whole line)
+     *
+     * @see https://www.gnu.org/software/libc/manual/html_node/Canonical-or-Not.html
      */
-    private $originalConfiguration;
+    public function disableCanonicalMode() : void;
 
-    public function __construct()
-    {
-        $this->getOriginalConfiguration();
-    }
+    /**
+     * Enable canonical input - read input by line
+     *
+     * @see https://www.gnu.org/software/libc/manual/html_node/Canonical-or-Not.html
+     */
+    public function enableCanonicalMode() : void;
 
-    public function getWidth(): int
-    {
-        return $this->width ?: $this->width = (int) exec('tput cols');
-    }
+    /**
+     * Is canonical mode enabled or not
+     */
+    public function isCanonicalMode() : bool;
 
-    public function getHeight(): int
-    {
-        return $this->height ?: $this->height = (int) exec('tput lines');
-    }
+    /**
+     * Check if the Input & Output streams are interactive. Eg - they are
+     * connected to a terminal.
+     *
+     * @return bool
+     */
+    public function isInteractive() : bool;
 
-    public function getDetails(): string
```

Comment:
I've dropped this completely. Basically this got the name of the terminal. But it was only used when the terminal was not a TTY to throw an exception. But if it's not a TTY - then it doesn't really have a name because it's either a pipe or a redirect. So I don't think we need this.